src/cli/serve/compile_config_stack.js

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import _ from 'lodash';
+
+import { readFileSync, writeFileSync, statSync, existsSync } from 'fs';
+import { resolve } from 'path';
+import { getConfigPath, getConfigDirectory } from '@kbn/utils';
+import { getConfigFromFiles } from '@kbn/config';
+
+const isNotEmpty = _.negate(_.isEmpty);
+const isNotNull = _.negate(_.isNull);
+
+/** @typedef {'es' | 'oblt' | 'security'} ServerlessProjectMode */
+/** @type {ServerlessProjectMode[]} */
+const VALID_SERVERLESS_PROJECT_MODE = ['es', 'oblt', 'security'];
+
+/**
+ * Collects paths to configurations to be included in the final configuration stack.
+ * @param {{configOverrides?: string[], devConfig?: boolean, dev?: boolean, serverless?: string | true}} options Options impacting the outgoing config list
+ * @returns List of paths to configurations to be merged, from left to right.
+ */
+export function compileConfigStack({ configOverrides, devConfig, dev, serverless }) {
+  const cliConfigs = configOverrides || [];
+  const envConfigs = getEnvConfigs();
+  const defaultConfig = getConfigPath();
+
+  let configs = [cliConfigs, envConfigs, [defaultConfig]].find(isNotEmpty);
+
+  if (dev && devConfig !== false) {
+    configs.push(resolveConfig('kibana.dev.yml'));
+  }
+
+  if (dev && serverless) {
+    writeProjectSwitcherConfig('serverless.recent.dev.yml', serverless);
+    configs.push(resolveConfig('serverless.recent.dev.yml'));
+  }
+
+  // Filter out all config paths that didn't exist
+  configs = configs.filter(isNotNull);
+
+  const serverlessMode = validateServerlessMode(serverless) || getServerlessModeFromCfg(configs);
+  if (serverlessMode) {
+    configs.unshift(resolveConfig(`serverless.${serverlessMode}.yml`));
+    configs.unshift(resolveConfig('serverless.yml'));
+
+    if (dev && devConfig !== false) {
+      configs.push(resolveConfig('serverless.dev.yml'));
+      configs.push(resolveConfig(`serverless.${serverlessMode}.dev.yml`));
+    }
```

Comment:
Spencer's original PR had support for these project-specific `.dev.yml` files which I think got lost in a later update:

https://github.com/elastic/kibana/blob/f48b742be10d1501383ba63f47777c9245f40a8b/config/README.md?plain=1#L7-L16

I don't have a strong feeling on this. Since `kibana.dev.yml` always wins, I don't think there's a functional need for the project-specific dev overrides. It's just more of an organizational question... If folks want to override a security-specific setting, for example, it may be clearer to have the ability to copy that into a designated file that they could easily map back to the original security serverless yml.

It's a dev-only feature, so if there's anyone who would find it useful and it doesn't cost us much to maintain, then I don't see any harm in keeping it.

Reply:
I'm easily swayed towards either behaviour, let me know which one you prefer (keep the specific configs, or only have `kibana.dev.yml`)

Reply:
> It's a dev-only feature, so if there's anyone who would find it useful and it doesn't cost us much to maintain, then I don't see any harm in keeping it.

My main concern is that a dev perceives a different behaviour just because they are using those `serverless.*dev.yml` files and forgot about them.

Having said that, I don't think we need to increase the scope of this PR. We can discuss it, and remove those later if needed.